File: bench/cpp/fft.cpp
// bench/cpp/fft.cpp — 迭代 radix-2 FFT（与 bench/myp/fft.myp 同算法）
// 输出：verify <频谱能量> / ms <毫秒>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <vector>

static const double PI = 3.141592653589793;

static double fft(int n, int iters) {
    std::vector<double> re(n), im(n);
    for (int it = 0; it < iters; ++it) {
        // 每次重置确定性有界信号（避免反复 FFT 导致溢出）
        for (int i = 0; i < n; ++i) {
            re[i] = std::sin((i % 100) * 0.1234567);
            im[i] = std::cos((i % 100) * 0.07654321);
        }
        // 位反转
        int j = 0;
        for (int i = 1; i < n; ++i) {
            int bit = n >> 1;
            while ((j & bit) != 0) { j ^= bit; bit >>= 1; }
            j ^= bit;
            if (i < j) { std::swap(re[i], re[j]); std::swap(im[i], im[j]); }
        }
        // 蝶形
        for (int len = 2; len <= n; len <<= 1) {
            double ang = -2.0 * PI / len;
            double wRe = std::cos(ang), wIm = std::sin(ang);
            for (int k = 0; k < n; k += len) {
                double curRe = 1.0, curIm = 0.0;
                for (int k2 = 0; k2 < len / 2; ++k2) {
                    int a = k + k2, b = k + k2 + len / 2;
                    double tre = re[b], tim = im[b];
                    double pRe = tre * curRe - tim * curIm;
                    double pIm = tre * curIm + tim * curRe;
                    re[b] = re[a] - pRe;
                    im[b] = im[a] - pIm;
                    re[a] += pRe;
                    im[a] += pIm;
                    double ncRe = curRe * wRe - curIm * wIm;
                    double ncIm = curRe * wIm + curIm * wRe;
                    curRe = ncRe;
                    curIm = ncIm;
                }
            }
        }
    }
    double e = 0.0;
    for (int i = 0; i < n; ++i) e += re[i] * re[i] + im[i] * im[i];
    return e;
}

int main() {
    const int n = 4096, iters = 800;
    auto t0 = std::chrono::steady_clock::now();
    double e = fft(n, iters);
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("verify %.10g\nms %.0f\n", e, ms);
    return 0;
}

File: bench/go/fft.go
// bench/go/fft.go — Go 版迭代 radix-2 FFT（与 bench/myp/fft.myp 同算法）
// verify <频谱能量>（浮点，容差对比）；与 C++/MYP 同算法。
package main

import (
	"fmt"
	"math"
	"time"
)

const pi = 3.141592653589793

func fft(n, iters int) float64 {
	re := make([]float64, n)
	im := make([]float64, n)
	for it := 0; it < iters; it++ {
		for i := 0; i < n; i++ {
			re[i] = math.Sin(float64(i%100) * 0.1234567)
			im[i] = math.Cos(float64(i%100) * 0.07654321)
		}
		j := 0
		for i := 1; i < n; i++ {
			bit := n >> 1
			for (j & bit) != 0 {
				j ^= bit
				bit >>= 1
			}
			j ^= bit
			if i < j {
				re[i], re[j] = re[j], re[i]
				im[i], im[j] = im[j], im[i]
			}
		}
		for length := 2; length <= n; length <<= 1 {
			ang := -2.0 * pi / float64(length)
			wRe := math.Cos(ang)
			wIm := math.Sin(ang)
			for k := 0; k < n; k += length {
				curRe, curIm := 1.0, 0.0
				for k2 := 0; k2 < length/2; k2++ {
					a := k + k2
					b := k + k2 + length/2
					tre := re[b]
					tim := im[b]
					pRe := tre*curRe - tim*curIm
					pIm := tre*curIm + tim*curRe
					re[b] = re[a] - pRe
					im[b] = im[a] - pIm
					re[a] += pRe
					im[a] += pIm
					ncRe := curRe*wRe - curIm*wIm
					ncIm := curRe*wIm + curIm*wRe
					curRe = ncRe
					curIm = ncIm
				}
			}
		}
	}
	e := 0.0
	for i := 0; i < n; i++ {
		e += re[i]*re[i] + im[i]*im[i]
	}
	return e
}

func main() {
	const n, iters = 4096, 800
	t0 := time.Now()
	e := fft(n, iters)
	ms := time.Since(t0).Milliseconds()
	fmt.Printf("verify %.10g\nms %d\n", e, ms)
}
